CMakeLists.txt

```diff
@@ -0,0 +1,172 @@
+cmake_minimum_required(VERSION 2.6)
+
+project(UriParser C)
+
+set(VERSION_STRING "0.8.6")
+IF(EXISTS "${CMAKE_SOURCE_DIR}/.git")
+    FIND_PROGRAM(GIT git)
+    IF(GIT)
+        EXECUTE_PROCESS(
+            WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+            COMMAND ${GIT} describe --tags
+            OUTPUT_VARIABLE GIT_VERSION  OUTPUT_STRIP_TRAILING_WHITESPACE
+        )
+        if(NOT VERSION_STRING)
+                EXECUTE_PROCESS(
+                    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+                    COMMAND ${GIT} rev-parse --short HEAD
+                    OUTPUT_VARIABLE GIT_VERSION  OUTPUT_STRIP_TRAILING_WHITESPACE
+                )
+        endif()
+        SET(VERSION_STRING ${GIT_VERSION})
+    ENDIF(GIT)
+ENDIF()
+MESSAGE("Version: ${VERSION_STRING}\n")
+set(URIPARSER_VERSION ${VERSION_STRING})
+
+option(BUILD_TESTS "Build demos." OFF)
+
+IF(MSVC)
+    # This option is to enable the /MP switch for Visual Studio 2005 and above compilers
+    OPTION(WIN32_USE_MP "Set to ON to build OpenSceneGraph with the /MP option (Visual Studio 2005 and above)." ON)
+    MARK_AS_ADVANCED(WIN32_USE_MP)
+    IF(WIN32_USE_MP)
+        SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /MP")
+    ENDIF(WIN32_USE_MP)
+ENDIF(MSVC)
+
+include(CMakePackageConfigHelpers)
+include(GNUInstallDirs OPTIONAL)
+include(GenerateExportHeader)
+include(CheckIncludeFile)
+include(CheckFunctionExists)
+
+check_include_file("wchar.h" HAVE_WCHAR_H)
+check_include_file("stdio.h" HAVE_STDIO_H)
+
+check_function_exists(wprintf HAVE_WPRINTF)
+
+set(CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake ${CMAKE_MODULE_PATH})
+
+option(ENABLE_SIZEDOWN  "enable size down: -DURI_SIZEDOWN" ON)
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -DURI_SIZEDOWN")
+
+option(DISABLE_CHAR "disable char: -DURI_NO_ANSI" OFF)
+if(DISABLE_CHAR)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -DURI_NO_ANSI")
+endif()
+
+option(DISABLE_WCHAR_T "disable wchar_t: -DURI_NO_UNICODE " OFF)
+if(DISABLE_WCHAR_T)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -DURI_NO_UNICODE")
+endif()
+
+if(DISABLE_WCHAR_T AND DISABLE_CHAR)
+   message(FATAL_ERROR "DISABLE_WCHAR_T and DISABLE_CHAR are mutually exclusive")
+endif()
+
+if(NOT (DISABLE_WCHAR_T OR DISABLE_CHAR) AND BUILD_TESTS)

```

Comment:
Let's also produce an error for the else case here saying that the requirements of the test suite are violated rather than just silently dropping test execution. Thanks!

Reply:
It is add by you:)